_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+types, our industry's code bases are still full of pointers to arrays. So this warning gives a
+mechanism to stard migrating away from pointers, _and then prevent backsliding_.
+
+## Performance implications
+
+Introducing memory safety into C++ codebases always comes with some performance degredation, because
+there's little you can check at compile time. So it means checking for bad program states when the
+programming is running. But these performance changes are not the kind you as a user will see or
+feel. But the kind that shows up on benchmarks and that software teams then have great a many
+[discussions about](https://groups.google.com/a/chromium.org/g/platform-architecture-dev/c/hdMBHEYZLMA).
+And these often end up blocking meaningful improvements to the safety of users and their data (for
+better or for worse).
+
+In bounds checking, the compiler can see the values involved in the comparisons, as they come from
+constants or other values related to the array. For instance when iterating through a span, stopping
+before the span's `length`, the compiler can see that every index into the span is less than its
+`length` fairly trivially. This usually means the compiler removes the bounds checks. So no runtime
+overhead, yay!
+
+But there's another type of overhead that comes not from the bounds checks in view types like
+`std::span` or `std::string_view`, but from the change to using those types at all.
+
+Here's a commit that converts a bunch of functions to use `base::StringPiece`, the Chromium string
+view type which is much like `std::string_view`:
+[https://chromium-review.googlesource.com/c/chromium/src/+/4575826](
+https://chromium-review.googlesource.com/c/chromium/src/+/4575826).
+
+The functions were previously overloaded to take `const std::string&` or `const char*`. The
+`const char*` overload presumes that:
+- the string being passed in is null terminated, and
+- the function should be consuming the full string up to the null character.
+
+These assumptions are easily wrong when working with view types that encode a bounded view of an
+array. Calling these functions requires converting that bounded string view into a `std::string`
+which involves generating a constructor, and quite possibly a heap allocation. But
+many callers instead convert their bounded string view into a `const char*` through `.data()`
+instead, which is a bug. Props to [@davidben](https://github.com/davidben) for noticing this common
+anti-pattern in Chromium and working to fix them.
+
+So it appears at first glance to be a win to change these functions to take a string view type, in
+this case `base::StringPiece`. For callers with a `std::string`, it will copy the pointer and length
+into the `base::StringPiece`. For callers with a bounded string view type, they don't need to throw
+away the bounds or heap allocate. Yet this created a pretty large binary size regression for doing

```

Comment:
compound verb needs hyphen: heap-allocate; or "allocate on the heap" or "dynamically allocate"